[GStreamer][WebRTC] Increase spec compliance for ICE candidate stats

philn opened the pull request:
#### 53ac924b3cfcbf0dc4237e4cba2ceead732d334c
<pre>
[GStreamer][WebRTC] Increase spec compliance for ICE candidate stats
<a href="https://bugs.webkit.org/show_bug.cgi?id=290451">https://bugs.webkit.org/show_bug.cgi?id=290451</a>

Reviewed by Xabier Rodriguez-Calvar.

Implement the new ICETransport-&gt;get_selected_candidate_pair vfunc in the Rice backend and fill
ICE candidate stats from GStreamer ICE candidate stats. This will allow us to unflag several layout
tests once the SDK is updated to GStreamer 1.28.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransport.cpp:
(populateCandidateStats):
(fillCredentials):
(riceCandidateToGst):
(webkitGstWebRTCIceTransportGetSelectedCandidatePair):
(webkit_gst_webrtc_ice_transport_class_init):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransportBackend.cpp:
(WebCore::candidateFromGstWebRTC):
(WebCore::GStreamerIceTransportBackend::selectedCandidatePairChanged):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::IceCandidateStats::IceCandidateStats):
(WebCore::iceCandidateType): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::toRTCIceComponent): Deleted.
(WebCore::toRTCIceProtocol): Deleted.
(WebCore::toRTCIceTcpCandidateType): Deleted.
(WebCore::toRTCIceCandidateType): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
(WebCore::toRTCIceComponent):
(WebCore::toRTCIceProtocol):
(WebCore::toRTCIceTcpCandidateType):
(WebCore::toRTCIceCandidateType):
* Source/WebCore/platform/graphics/gstreamer/GUniquePtrGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/304217@main">https://commits.webkit.org/304217@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b5d5bf251c86b6220284c55323b8257434ab0bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46189 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142451 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136812 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7206 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5647 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83955 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5469 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3047 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39108 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145151 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7036 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/39683 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111840 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28374 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5307 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117235 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7085 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6858 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70657 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/7093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransport.cpp:
```diff
@@ -234,16 +240,25 @@ static void populateCandidateStats(const RiceCandidate* candidate, GstWebRTCICEC
 {
     GUniquePtr<RiceCredentials> credentials(isLocal ? rice_stream_get_local_credentials(stream.get()) : rice_stream_get_remote_credentials(stream.get()));
-    // rice_credentials_get_ufrag_bytes(credentials.get(), stats->);
+    std::array<char, 256> bytes;
+    auto size = rice_credentials_get_ufrag_bytes(credentials.get(), bytes.data());
```
Strongly encourage:

```suggestion
    std::array<char, 256> bytes;
    auto size = rice_credentials_get_ufrag_bytes(credentials.get(), bytes.data());
    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(size <= 256);
```

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransport.cpp:
```diff
@@ -277,10 +292,51 @@ bool webkitGstWebRTCIceTransportGetSelectedPair(WebKitGstIceTransport* transport
+    if (!candidate)
+        return gstCandidate;
+
+    /* FIXME: sdpMid, sdpMLineIndex */
```
//

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransportBackend.cpp:
```diff
@@ -150,11 +151,69 @@ void GStreamerIceTransportBackend::gatheringStateChanged() const
 }
 
+#if GST_CHECK_VERSION(1, 27, 0)
+static Ref<RTCIceCandidate> candidateFromGstWebRTC(const GstWebRTCICECandidate* c)
```
name the parameter as candidate, please

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransportBackend.cpp:
```diff
@@ -150,11 +151,69 @@ void GStreamerIceTransportBackend::gatheringStateChanged() const
+            break;
+        };
+
+        auto relatedAddress = StringView::fromLatin1(GST_WEBRTC_ICE_CANDIDATE_STATS_RELATED_ADDRESS(c->stats));
```
I would use CStringView.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
```diff
@@ -49,6 +52,47 @@
+    return RTCIceProtocol::Tcp;
+}
+
+inline std::optional<RTCIceTcpCandidateType> toRTCIceTcpCandidateType(StringView type)
```
I would use CStringView

philn replied:
tcptype in `parseIceCandidateSDP()` is a String, so I don't think a CStringView makes sense here.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
```diff
@@ -49,6 +52,47 @@
+    return RTCIceTcpCandidateType::So;
+}
+
+inline std::optional<RTCIceCandidateType> toRTCIceCandidateType(StringView type)
```
I would use CStringView

philn replied:
type in parseIceCandidateSDP() is a String, so I don't think a CStringView makes sense here.

philn committed:
[GStreamer][WebRTC] Increase spec compliance for ICE candidate stats
https://bugs.webkit.org/show_bug.cgi?id=290451

Reviewed by Xabier Rodriguez-Calvar.

Implement the new ICETransport->get_selected_candidate_pair vfunc in the Rice backend and fill
ICE candidate stats from GStreamer ICE candidate stats. This will allow us to unflag several layout
tests once the SDK is updated to GStreamer 1.28.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransport.cpp:
(populateCandidateStats):
(fillCredentials):
(riceCandidateToGst):
(webkitGstWebRTCIceTransportGetSelectedCandidatePair):
(webkit_gst_webrtc_ice_transport_class_init):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceTransportBackend.cpp:
(WebCore::candidateFromGstWebRTC):
(WebCore::GStreamerIceTransportBackend::selectedCandidatePairChanged):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::IceCandidateStats::IceCandidateStats):
(WebCore::iceCandidateType): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::toRTCIceComponent): Deleted.
(WebCore::toRTCIceProtocol): Deleted.
(WebCore::toRTCIceTcpCandidateType): Deleted.
(WebCore::toRTCIceCandidateType): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:
(WebCore::toRTCIceComponent):
(WebCore::toRTCIceProtocol):
(WebCore::toRTCIceTcpCandidateType):
(WebCore::toRTCIceCandidateType):
* Source/WebCore/platform/graphics/gstreamer/GUniquePtrGStreamer.h:

Canonical link: https://commits.webkit.org/304217@main
